src/main/kotlin/resource/UserResource.kt

```diff
@@ -34,6 +57,76 @@ fun Route.user(userService: UserService) {
                 call.respond("You are missing one or multiple parameters")
             }
         }
+
+        webSocket("/balance") {
+            try {
+                val user = TokenManager.verifyTokenAndRetrieveUser(call.parameters)
+                val balanceManager = userService.buildBalanceManager(user)
+                val balanceIncreaseRateManager = BalanceIncreaseRateManager(user)
+
+                SessionSemaphore.acquireBalanceSession(user)
+
+                try {
+                    while (true) {
+                        balanceIncreaseRateManager.increaseBalanceBasedOnIncreaseRate(balanceManager)
+
+                        val currentBalance = balanceManager.retrieveCurrentBalance()
+                        outgoing.send(Frame.Text(currentBalance.toString()))
+
+                        delay(TimeUnit.SECONDS.toMillis(Scheduler.BalanceIncreaseTimeoutInSeconds))
+                    }
+                } catch (exception: Exception) {
+                    close(CloseReason(CloseReason.Codes.UNEXPECTED_CONDITION, message = WebSocketClosedByExceptionMessage))
+                } finally {
+                    balanceManager.syncCurrentBalanceToDatabase()
+                    SessionSemaphore.releaseBalanceSession(user)
+                }
+            } catch (exception: TokenExpiredException) {
```

Comment:
Merge with exception below.

Reply:
Turns out one cannot do multi-exceptions in Kotlin, but I shortened it with some pattern matching.